### ISSUE TITLE
"*_(cosplay)" creates broken tags

### ISSUE THREAD

**Issue by kittey** [r888888888/danbooru#3307]
The user Allynay [discovered](https://danbooru.donmai.us/forum_posts/136153) in his BigQuery dataset that there are some tags on posts that aren’t in the tag table.

* [idunn_(p&d)](https://danbooru.donmai.us/wiki_pages/show_or_new?title=idunn_%28p%26d%29) on [post #1858377](https://danbooru.donmai.us/posts/1858377)
* [anila_(granblue_fantasy))](https://danbooru.donmai.us/wiki_pages/show_or_new?title=anila_%28granblue_fantasy%29%29) on [post #2351218](https://danbooru.donmai.us/posts/2351218)
* [kishii_mika](https://danbooru.donmai.us/wiki_pages/show_or_new?title=kishii_mika) on [post #2540508](https://danbooru.donmai.us/posts/2540508)

All tags have no count on the post pages and were apparently added when their corresponding *_(cosplay) tags were added.

Possibly related to issue #2868.

**Comment by evazion** [comment]
https://github.com/r888888888/danbooru/blob/71c38208388cd8eb5d1b34dc8ce1a9deb84d4283/app/models/post.rb#L684-L698

The bug is that automatic tags are added to the tag string after tags are created (`Tag.find_or_create_by_name`). Technically this could also happen with normal automatic tags (e.g. `tagme`, `highres`) in a fresh install where these tags don't exist in the tags table yet.

**Comment by evazion** [comment]
This turns out to be a little tricky, due to the way `Tag.find_or_create_by_name` works. We want tags to be created in the final step, after aliases/implications/automatic tags are applied. But this isn't possible because `Tag.find_or_create_by_name` also handles the `gen:`/`char:`/`copy:`/`art:` tag type prefixes, and these prefixes need to be stripped from tag names early on, \*before\* aliases/implications/automatic tags are applied.

Basically, the `gen:`/`char:`/`copy:`/`art:` handling needs to be moved from `Tag.find_or_create_by_name` to `Post#apply_post_metatags` in order to make this work.

**Comment by evazion** [comment]
da170aa made it so that if you add e.g. `gen:stockings` to a post then the `stockings` -> `thighhighs` alias won't be applied and the `stockings` tag will be incorrectly added to the post. In this way one can add unsearchable tags to posts. This is why tag type prefixes need to be handled before aliases.

**Comment by r888888888** [comment]
Hmm, this gets into the situation of what the expected behavior of `gen:stockings` should be assuming `stockings -> thighhighs` alias exists. Should it change the type of `stockings` or should it try to be smart and apply the change to `thighhighs`.

**Comment by r888888888** [comment]
Fixed
